src/v/migrations/rbac_migrator.cc

```diff
@@ -42,7 +42,17 @@ ss::future<> rbac_migrator::do_mutate() {
     auto err = co_await _controller.get_security_frontend().local().create_role(
       role_name, std::move(role), model::timeout_clock::now() + 5s);
 
-    if (err) {
+    if (err == cluster::errc::role_exists) {
```

Comment:
Is it possible that in-between more users were created? Are migrations guaranteed to run before any other operations on the cluster can happen?

We should probably document explicitly (if not done already) that migrations must be idempotent.

Reply:
> Is it possible that in-between more users were created?

Yes, and that's fine. In that case, the migration will appear to have happened at the original time (before the users were added). This role is created as a one-time convenience to showcase roles and it won't be kept up-to-date with users added after the migration.

> Are migrations guaranteed to run before any other operations on the cluster can happen?

The logical version of the cluster is only incremented after the migration and the RBAC feature is only activated after the migration, so other RBAC operations guarded by the RBAC feature flag won't happen. But otherwise, operations aren't blocked by the migration as far as I can tell.

> We should probably document explicitly (if not done already) that migrations must be idempotent.

I now noticed that there was already a hint about migrations executing *at least once* ([here](https://github.com/redpanda-data/redpanda/blob/465118e4672c15bab1e19c86d4f886cd74e95742/src/v/migrations/feature_migrator.h#L21-L28)), but I've now added one more comment to state that `do_mutate` should be idempotent: https://github.com/redpanda-data/redpanda/pull/17800/commits/33d857fd1c38e5d727b41d4100f0117ca7bee1b5

Reply:
> and it won't be kept up-to-date with users added after the migration.

wondering why are we bothering to do it at all then :) wouldn't it just confuse users more?